libcanard/canard.h

```diff
@@ -1,528 +1,587 @@
+///
+/// The DSDL extension is trivial and there is not much to document -- please refer to its header file for details.
+///
+/// This transport layer implementation consists of two components: the transmission (TX) pipeline and the
+/// reception (RX) pipeline.
+///
+/// The TX and RX pipelines are completely independent from each other except that they both rely on the same
+/// dynamic memory manager. The TX pipeline uses the dynamic memory to store outgoing CAN frames in the prioritized
+/// transmission queue. The RX pipeline uses the dynamic memory to store contiguous payload buffers for received
+/// transfers and for keeping the transfer reassembly state machine data. The exact memory consumption model is defined
+/// for both pipelines, so it is possible to statically determine the minimum size of the dynamic memory pool required
+/// to guarantee that a given application will never encounter an out-of-memory error at runtime.
+///
+/// Much like with dynamic memory, the time complexity of every API function is well-characterized, allowing the
+/// application to guarantee predictable real-time performance.
+///
+/// The TX pipeline is managed with the help of three API functions. When the application needs to emit a transfer,
+/// it invokes canardTxPush(). The function splits the transfer into CAN frames and stores them into the prioritized
+/// transmission queue. The application then picks the CAN frames from the queue one-by-one by calling canardTxPeek()
+/// followed by canardTxPop() -- the former allows the application to look at the frame and the latter tells the library
+/// that the frame shall be removed from the queue. The returned frames need to be deallocated by the application.
+///
+/// The RX pipeline is managed with the help of three API functions. The main function canardRxAccept() takes a
+/// received CAN frame and updates the appropriate transfer reassembly state machine. The functions canardRxSubscribe()
+/// and its counterpart canardRxUnsubscribe() instruct the library which transfers should be received (by default, all
+/// transfers are ignored); also the subscription function specifies vital transfer reassembly parameters such as the
+/// maximum payload size (i.e., the maximum size of a serialized representation of a DSDL object) and the transfer-ID
+/// timeout. Transfers that carry more payload than the configured maximum per subscription are truncated following the
+/// Implicit Truncation Rule (ITR) defined by the UAVCAN Specification -- the rule is implemented to facilitate
+/// backward-compatible DSDL data type extensibility.
+///
+/// The library supports a practically unlimited number of redundant transports.
+///
+/// The library is not thread-safe: if used in a concurrent environment, it is the responsibility of the application
+/// to provide adequate synchronization.
+///
+/// The library is purely reactive: it does not perform any background processing and does not require periodic
```

Comment:
so, you have to implement a driver using interrupts then if you use libcanard?

Reply:
No, it is not necessary to use interrupts. You can poll your CAN controller if you want. The point is that the library will not do that for you; it is your responsibility to feed incoming frames into it and collect reassembled transfers at the other end.

Reply:
I get it now. Sorry. was still thinking this was a layered architecture but now I see it is not.